adapter: enable QUnit configuration passthrough

mduvall opened the pull request:
### What's in here

This PR adds the ability to pass through [QUnit `config` values](https://api.qunitjs.com/QUnit.config/) to the `window.QUnit` instance. It preserves the `autostart` behavior and allows clients to override by writing into `client.qunit` as the [`karma-mocha` allows](https://github.com/karma-runner/karma-mocha).

The original intent for this PR is to enable clients to do easy module filtering, @marcins took a stab at this in https://github.com/karma-runner/karma-qunit/pull/29, but unfortunately it appears that `filter` is an undocumented configuration value for `window.QUnit.config` and I would rather insist on using the public API here for stability. As such, I do not want to be pinned to specific adapter keys at the top level that tie themselves to a key on `window.QUnit.config` (in the case of #29 this is `filter`), but I want to enable a passthrough of all values onto the `window.QUnit` instance so that future upgrades to QUnit behind the scenes are transparent to clients.

An example configuration would look like:

```js
module.exports = function(config) {
  config.set({
    ...
    client: {
      qunit: {
        moduleFilter: '<pattern>',
        ...
      }
      ...
    }
  });
};
```

Please let me know if this is amenable and any changes that need to be made for this to be merged. Thanks!

zzo commented:
can you update the README.md please with this info

marcins commented:
@zzo I'm guessing it's to minimise changes in the wrapper, and keep most of the logic in the adapter. Although the adapter probably shouldn't know about the `karma` object either?

@mduvall - an alternative approach might be to make the function in the adapter `configureQUnit(QUnit, config, defaultConfig)` and call it in the adapter like:

```js
if (window.__karma__.config && window.__karma__.config.qunit) {
    configureQUnit(window.QUnit, wndow.__karma__.config.qunit, { autostart: false });
}
```

Should probably also add unit tests for anything added to the adapter.

mduvall commented:
@zzo what @marcins said, I was just pattern matching off of [karma-mocha](https://github.com/karma-runner/karma-mocha/blob/master/src/adapter.js)'s adapter where it accesses the `karma` object. I'm not partial to either approach, I figured keeping the construction of the config. in relation to what is provided in the `karma` object seemed nice for locality / proximity to using code.

mduvall commented:
Also will update README, any suggestions on best way to provide an actual runner / if that infra is set up in any other repo? It currently uses a `MockRunner()` and seems to test the integration with the start hook (see [here](https://github.com/karma-runner/karma-qunit/blob/master/test/adapter.spec.js)).

mduvall commented:
Hello @zzo,

I have added the requested changes for the README, I have also added unit tests to exercise the logic within the configuration extraction. Let me know if there is anything else I can add.

mduvall commented:
Also adding @Dignifiedquire, who has cut the latest release. Do you mind taking a look at this?

zzo commented:
lgtm but i'll wait for @Dignifiedquire to comment before doing anything rash :)

Dignifiedquire commented:
Thanks a lot this looks great, just can you change your commit messages to follow our [convention](http://karma-runner.github.io/0.13/dev/git-commit-msg.html) please?

mduvall commented:
@Dignifiedquire Ah yes, I am sorry I missed that. I have updated the commit messages to reflect the conventions, let me know if there is anything else.

mduvall commented:
@Dignifiedquire Is this good to merge? Is it also possible to cut a new release after this as well? (happy to open PR for that if so)

Dignifiedquire commented:
Looks good to me @mduvall thanks :octocat:

Dignifiedquire commented:
Published as `0.1.6`

mduvall commented:
Great, thanks so much! :cake:

Dignifiedquire commented:
Getting this failure with the latest version, any ideas? https://travis-ci.org/karma-runner/karma/jobs/86863029

zzo commented:
fun - no ideas will try to replicate that locally

On Thu, Oct 22, 2015 at 10:05 AM, Friedel Ziegelmayer <
notifications@github.com> wrote:

> Getting this failure with the latest version, any ideas?
> https://travis-ci.org/karma-runner/karma/jobs/86863029
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/karma-runner/karma-qunit/pull/30#issuecomment-150292357>
> .
>

zzo commented:
autostart is not getting set properly - qunit thinks autostart is true

zzo commented:
ah the problem is this in adapter.js (line 6):

  if (!karma.config || !karma.config.qunit) {
    return {}
  }

if there is no specific qunit config in karma.conf.js then the default 'autostart:false' gets thrown away!

mduvall commented:
Yep, just opened: https://github.com/karma-runner/karma-qunit/pull/31
